### PR TITLE
Refactor of Google Adapter

### PR DESCRIPTION
**Pull request by garypretty**
Aligned with the current preview of the Alexa adapter.  

* Supports all Google helper intents, such as Carousel, List signin etc.
* Supports both DialogFlow / Conversation Webhook - now both aligned in terms of functionlity (had been diverging over time)
* Improved BF activity mapping - incl. ConversationUpdate.
* Removed support for Web API, as with Alexa.
* Updated sample

### CODE REVIEW

File: libraries/Bot.Builder.Community.Adapters.Alexa.Core/Helpers/AlexaMarkdownToPlaintextRenderer.cs [hunk 1]
```diff
@@ -1,70 +1,70 @@
...
-            public override string Table(string header, string body) => string.Empty;
-            public override string TableCell(string content, TableCellFlags flags) => string.Empty;
-            public override string TableRow(string content) => string.Empty;
-            public override string Text(string text) => text.TrimEnd('.');
-            public override string Preprocess(string text) => text.Trim();
-            public override string Postprocess(string text) => text.Trim();
-        }
-    }
-}
+﻿using Microsoft.MarkedNet;
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Alexa.Core/Helpers/AlexaMarkdownToPlaintextRenderer.cs:1]: Newlines or something I assume.

Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Alexa.Core/Helpers/AlexaMarkdownToPlaintextRenderer.cs:1]: I think this is something to do with the merge VS did, but I will check.

File: libraries/Bot.Builder.Community.Adapters.Google.Core/Attachments/AttachmentHelper.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                    case GoogleAttachmentContentTypes.ConfirmationIntent:
+                        Convert<ConfirmationIntent>(attachment);
+                        break;
+                }
+            }
+        }
+
+        private static void Convert<T>(Attachment attachment)
+        {
+            // Alexa-skills-dotnet has a custom JsonConverter that converts ICards and IDirective to their correct type so we don't need to do that.
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google.Core/Attachments/AttachmentHelper.cs:68]: Minor: Alexa noted here

File: libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/Extensions.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using Bot.Builder.Community.Adapters.Google.Core.Model;
+using Bot.Builder.Community.Adapters.Google.Core.Model.Request;
+using Newtonsoft.Json.Linq;
+
+namespace Bot.Builder.Community.Adapters.Google.Core.Helpers
+{
+    public static class Extensions
+    {
+        public static void EnsureUniqueUserIdInUserStorage(this ConversationRequest conversationRequest)
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/Extensions.cs:10]: These are interesting, where is the storage here? Is it stored in Google?

File: libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/GoogleHelperIntentFactory.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        public static ListIntent CreateListIntent(string title, List<OptionItem> items)
+        {
+            return new ListIntent()
+            {
+                InputValueData = new ListOptionIntentInputValueData()
+                {
+                    ListSelect = new OptionIntentSelect()
+                    {
+                        Title = "This is the list title",
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/GoogleHelperIntentFactory.cs:18]: Set title here.

File: libraries/Bot.Builder.Community.Adapters.Google/Bot.Builder.Community.Adapters.Google.csproj [hunk 5]
```diff
@@ -15,9 +15,15 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <Compile Remove="Directives\**" />
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google/Bot.Builder.Community.Adapters.Google.csproj:18]: Same folder or old stuff from when you were splitting the code?

File: libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapter.cs [hunk 6]
```diff
@@ -1,454 +1,155 @@
...
-            throw new NotImplementedException();
+            return Task.FromResult(new ResourceResponse[0]);
         }
 
-        public override Task DeleteActivityAsync(ITurnContext turnContext, ConversationReference reference, CancellationToken cancellationToken)
+        private async Task<TurnContextEx> CreateContextAndRunPipelineAsync(IBot bot, CancellationToken cancellationToken, Activity activity)
         {
-            throw new NotImplementedException();
+            var context = new TurnContextEx(this, activity);
+            context.TurnState.Add("GoogleUserId", activity.From.Id);
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapter.cs:150]: needed or for back-compat?

File: libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapterOptions.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        public GoogleWebhookType WebhookType { get; set; } = GoogleWebhookType.Conversation;
+
+        public bool ShouldEndSessionByDefault { get; set; } = true;
+
+        public string ActionInvocationName { get; set; }
+
+        public string ActionProjectId { get; set; }
+
+        public bool ValidateIncomingRequests { get; set; }
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapterOptions.cs:19]: May be good to default this to true.

File: libraries/Bot.Builder.Community.Adapters.Google/TurnContextEx.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+using System.Collections.Generic;
+
+namespace Bot.Builder.Community.Adapters.Google
+{
+    internal class TurnContextEx : TurnContext
```
Comment (NickEricson) [inline libraries/Bot.Builder.Community.Adapters.Google/TurnContextEx.cs:7]: Needed, or can we use 4.8 expectReplies?

File: samples/Google Adapter Sample/Google Adapter Sample.csproj [hunk 9]
```diff
@@ -1,33 +1,25 @@
...
-      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
-    </Content>
+    <PackageReference Include="Bot.Builder.Community.Adapters.Google" Version="4.6.4-beta0033" />
+    <PackageReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.8.0" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\libraries\Bot.Builder.Community.Adapters.Google.Core\Bot.Builder.Community.Adapters.Google.Core.csproj" />
+    <ProjectReference Include="..\..\libraries\Bot.Builder.Community.Adapters.Google\Bot.Builder.Community.Adapters.Google.csproj" />
```
Comment (NickEricson) [inline samples/Google Adapter Sample/Google Adapter Sample.csproj:16]: Has project and package reference both.

File: samples/Google Adapter Sample/Startup.cs [hunk 10]
```diff
@@ -1,100 +1,67 @@
...
-
-                ILogger logger = _loggerFactory.CreateLogger<GoogleAdapterSampleBot>();
-
-                options.OnTurnError = async (context, exception) =>
+                return new GoogleAdapterOptions()
                 {
-                    logger.LogError($"Exception caught : {exception}");
-                    await context.SendActivityAsync("Sorry, it looks like something went wrong.");
+                    ActionInvocationName = "dialogflow adapter",
+                    //ActionProjectId = "adapter-helper-53294",
```
Comment (NickEricson) [inline samples/Google Adapter Sample/Startup.cs:36]: want this here?

File: libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/Extensions.cs (hunk 3 above)
Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/Extensions.cs:10]: Yeah, Google deprecated the creation of anonymous user IDs - see https://www.eisbahn.jp/yoichiro/2018/06/deprecated_anonymous_id_en.html

File: libraries/Bot.Builder.Community.Adapters.Google/Bot.Builder.Community.Adapters.Google.csproj (hunk 5 above)
Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Google/Bot.Builder.Community.Adapters.Google.csproj:18]: Removed

File: libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/GoogleHelperIntentFactory.cs (hunk 4 above)
Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Google.Core/Helpers/GoogleHelperIntentFactory.cs:18]: Fixed. Thanks.

File: libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapter.cs (hunk 6 above)
Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapter.cs:150]: Removed this last night :) Thanks.

File: libraries/Bot.Builder.Community.Adapters.Google/TurnContextEx.cs (hunk 8 above)
Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Google/TurnContextEx.cs:7]: We need this for the adapter scenario.  Expect replies will work via ABS if you are pushing an Activity through the BF adapter.  We can look to see if there is any consolidation that can be made in the near future though.

File: libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapterOptions.cs (hunk 7 above)
Comment (garypretty) [reply libraries/Bot.Builder.Community.Adapters.Google/GoogleAdapterOptions.cs:19]: Done. Good idea.

File: samples/Google Adapter Sample/Google Adapter Sample.csproj (hunk 9 above)
Comment (garypretty) [reply samples/Google Adapter Sample/Google Adapter Sample.csproj:16]: Yes, this is temporary until I get the updated package onto NuGet.

File: samples/Google Adapter Sample/Startup.cs (hunk 10 above)
Comment (garypretty) [reply samples/Google Adapter Sample/Startup.cs:36]: Amended.

**Pull request merged**